src/Io/RequestHeaderParser.php

```diff
@@ -24,20 +25,48 @@ class RequestHeaderParser extends EventEmitter
 {
     private $maxSize = 8192;
 
+    /**
+     * @var LoopInterface
+     */
+    private $loop;
+
     /** @var Clock */
     private $clock;
 
-    public function __construct(Clock $clock)
+    /**
+     * @var float
+     */
+    private $idleConnectionTimeout;
+
+    /**
+     * @param LoopInterface $loop
+     * @param float $idleConnectionTimeout
+     */
+    public function __construct(LoopInterface $loop, Clock $clock, $idleConnectionTimeout)
     {
+        $this->loop = $loop;
         $this->clock = $clock;
+        $this->idleConnectionTimeout = $idleConnectionTimeout;
     }
 
     public function handle(ConnectionInterface $conn)
     {
+        $loop = $this->loop;
+        $idleConnectionTimeout = $this->idleConnectionTimeout;
+        $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+            $conn->close();
+        });
+        $conn->on('close', function () use ($loop, &$timer) {
+            $loop->cancelTimer($timer);
+        });
         $buffer = '';
         $maxSize = $this->maxSize;
         $that = $this;
-        $conn->on('data', $fn = function ($data) use (&$buffer, &$fn, $conn, $maxSize, $that) {
+        $conn->on('data', $fn = function ($data) use (&$buffer, &$fn, $conn, $maxSize, $that, $loop, &$timer, $idleConnectionTimeout) {
+            $loop->cancelTimer($timer);
+            $timer = $loop->addTimer($idleConnectionTimeout, function () use ($conn) {
+                $conn->close();
+            });
```

Comment:
Not sure how relevant this is, but considering this is on the hot path I'd like to optimize for the happy path here: Remove timer on `data` and only restart if request is incomplete (unlikely). At the moment, this would cancel + add + cancel.

Reply:
Done